### ISSUE TITLE
Fix supporting encrypted configuration with integer keys

### ISSUE THREAD

**Issue by fabianof** [rails/rails#51215]
We should accept hashes with integer keys when deep transforming encrypted configuration. Example:

```
{1 => 1}
```

Before https://github.com/rails/rails/pull/42358 :

```
> deep_transform(hash)

{1=>1}
```

After https://github.com/rails/rails/pull/42358 :

```
> deep_transform(hash)

ruby/gems/3.3.0/gems/activesupport-7.1.3.2/lib/active_support/ordered_options.rb:38:in `[]=': undefined method `to_sym' for an instance of Integer (NoMethodError)

      super(key.to_sym, value)
               ^^^^^^^
Did you mean?  to_s

```

### System configuration
**Rails version**:
7.1.3.2

**Ruby version**:
3.0.2

**Comment by rafaelfranca** [comment]
When does an integer as credential name makes sense?

**Comment by fabianof** [comment]
Thanks for asking, Rafael.

Following is a real use case from one of my applications (using fake credentials), in which I have a "year => spreadsheet ID" mapping.

```
google:
  credentials: config/credentials/google/development.json
  spreadsheet_id:
    2021: 81ab723369a5f16cb68d22f67089b4754befbb32a91b
    2022: ede01aa536d886affa5a4795ff76766413f1a5b062f0
    2023: e0b894384abdcbfb7549ab23907435668f7b48c7ff12
    2024: a59545b56a896e821dcf8b8e6315a68c412bbbff531c
```

With that in place I've been used to do this for a long time:

```
Rails.application.credentials[:google][:spreadsheet_id][month.year]
```

Without accepting integer keys, I have to worry about parsing month.year to string (for example), which is not the behavior I would expect from a hash.

**Comment by zzak** [comment]
I looked into this, and came to the conclusion that this is just not supported.

See this branch as an example of why: https://github.com/rails/rails/compare/main...zzak:rails:51215

Basically, encrypted config uses [OrderedOptions](https://edgeapi.rubyonrails.org/classes/ActiveSupport/OrderedOptions.html) under the hood, it is designed to support method call syntax for digging into the hash.

The examples for [EncryptedConfiguration](https://edgeapi.rubyonrails.org/classes/ActiveSupport/EncryptedConfiguration.html) demonstrate this.

However, in Ruby integers are now able to be used for method definitions or calls.

```
syntax error, unexpected integer literal (SyntaxError)
def 1
    ^
```

This also doesn't work:

```
Thing.42
```

Since Encrypted Configuration is intended to support dynamic method lookup, we cannot have it both ways.
In other words, I don't think this a bug.

If you can think of a way to support your use-case without breaking backwards compatibility please open a PR.

**Comment by fabianof** [comment]
Thanks for your time looking into this, @zzak.

If there's no way to support this use-case, perhaps we should update [EncryptedConfiguration](https://edgeapi.rubyonrails.org/classes/ActiveSupport/EncryptedConfiguration.html) documentation accordingly, as it currently states that

> Values can be accessed via Hash methods, such as fetch and dig

Based on that it should accept integer keys too.

Also we can take this example from documentation:

> my_config.fetch(:foo)

From which we can infer that we can do this as well:

> my_config.fetch(1)

Which is acceptable by Ruby Hash:

```
$> {1 => 2}.fetch(1)
=> 2
```

**Comment by zzak** [comment]
@fabianof I don't think it's good API design to only allow certain methods be supported like this, and I've started working on a PR to raise an error pointing to the affected key in your case.

As a workaround, you can use string keys, and still access your data through `fetch` or `[]` but the method lookup syntax will not work (the Ruby interpreter will raise a syntax error, which is the best case scenario imo).
